diff --git a/executer/user/mcexec.c b/executer/user/mcexec.c
index 2d0d50e..9f109aa 100644
--- a/executer/user/mcexec.c
+++ b/executer/user/mcexec.c
@@ -470,7 +470,7 @@ struct program_load_desc *load_interp(struct program_load_desc *desc0, FILE *fp)
 			j++;
 		}
 	}
-	desc->num_sections = j;
+	desc->num_sections = -1; // Test for num_secionts check Issue#1011
 
 	desc->entry = hdr.e_entry;
 	desc->interp_align = align;
