# force_shutdown.patch COPYRIGHT FUJITSU LIMITED 2018
diff --git a/executer/kernel/mcctrl/driver.c b/executer/kernel/mcctrl/driver.c
index 1588f63..9d91154 100644
--- a/executer/kernel/mcctrl/driver.c
+++ b/executer/kernel/mcctrl/driver.c
@@ -171,12 +171,12 @@ int mcctrl_os_shutdown_notifier(int os_index)
 {
 	if (os[os_index]) {
 		/* Wait for os running */
-		if (ihk_os_wait_for_status(os[os_index], IHK_OS_STATUS_RUNNING, 0, 200) != 0) {
-			printk("IHK: OS does not become RUNNING in shutdown. Force shutdown.\n");
+//		if (ihk_os_wait_for_status(os[os_index], IHK_OS_STATUS_RUNNING, 0, 200) != 0) {
+//			printk("IHK: OS does not become RUNNING in shutdown. Force shutdown.\n");
 			/* send nmi to force shutdown */
 			ihk_os_send_nmi(os[os_index], 3);
 			mdelay(200);
-		}
+//		}
 
 		pager_cleanup();
 		sysfsm_cleanup(os[os_index]);
